File: packages/js/tflite-exporter/src/quantization/quantizer.ts
import { Graph, Node, Tensor, Attribute } from '@onnx9000/core';
import { QuantizationParameters } from '../flatbuffer/schema';

export interface QuantizationContext {
  mode: 'int8' | 'fp16' | 'none';
  // 329. Allow custom tflite quantization schema extensions manually via JS API arguments.
  customQuantizationMap?: Record<string, TensorQuantization>;
}

export interface TensorQuantization {
  min: number[];
  max: number[];
  scale: number[];
  zeroPoint: number[];
  quantizedDimension: number;
}

export class Quantizer {
  private graph: Graph;
  private ctx: QuantizationContext;
  private quantizationMap = new Map<string, TensorQuantization>();

  constructor(graph: Graph, ctx: QuantizationContext) {
    this.graph = graph;
    this.ctx = ctx;

    if (ctx.customQuantizationMap) {
      for (const [name, q] of Object.entries(ctx.customQuantizationMap)) {
        this.quantizationMap.set(name, q);
      }
    }
  }

  public getQuantizationOffset(builder: any, tensor: Tensor): number {
    const q = this.quantizationMap.get(tensor.name);
    if (!q) return 0;

    // 231. Encode QuantizationParameters table natively.
    let minOffset = 0,
      maxOffset = 0,
      scaleOffset = 0,
      zpOffset = 0;

    if (q.min.length > 0) {
      builder.startVector(4, q.min.length, 4);
      for (let i = q.min.length - 1; i >= 0; i--) builder.addFloat32(q.min[i]!);
      minOffset = builder.endVector(q.min.length);
    }
    if (q.max.length > 0) {
      builder.startVector(4, q.max.length, 4);
      for (let i = q.max.length - 1; i >= 0; i--) builder.addFloat32(q.max[i]!);
      maxOffset = builder.endVector(q.max.length);
    }

    // 232. Support scale (Float array) definitions.
    if (q.scale.length > 0) {
      builder.startVector(4, q.scale.length, 4);
      for (let i = q.scale.length - 1; i >= 0; i--) builder.addFloat32(q.scale[i]!);
      scaleOffset = builder.endVector(q.scale.length);
    }

    // 233. Support zero_point (Int64 array) definitions.
    if (q.zeroPoint.length > 0) {
      builder.startVector(8, q.zeroPoint.length, 8);
      for (let i = q.zeroPoint.length - 1; i >= 0; i--) {
        // Write 64-bit int. We'll write lower and upper 32 bits
        const zp = q.zeroPoint[i]!;
        const lower = zp | 0;
        const upper = Math.floor(zp / 4294967296);
        builder.writeInt32(upper);
        builder.writeInt32(lower);
      }
      zpOffset = builder.endVector(q.zeroPoint.length);
    }

    // 238. Extract quantized_dimension correctly for Per-Channel ops.
    return QuantizationParameters.create(
      builder,
      minOffset,
      maxOffset,
      scaleOffset,
      zpOffset,
      0, // details_type
      0, // details_offset
      q.quantizedDimension,
    );
  }

  public quantize(): void {
    if (this.ctx.mode === 'none') return;

    if (this.ctx.mode === 'fp16') {
      this.quantizeFP16();
    } else if (this.ctx.mode === 'int8') {
      this.quantizeINT8();
    }
  }

  private quantizeFP16(): void {
    // 241. Downcast FLOAT32 FlatBuffer arrays entirely to FLOAT16 bytes explicitly for FP16 models.
    for (const [name, tensor] of Object.entries(this.graph.tensors)) {
      if (
        tensor.dtype === 'float32' &&
        tensor.isInitializer &&
        tensor.data instanceof Float32Array
      ) {
        tensor.dtype = 'float16';
        // Convert f32 array to f16 bytes
        tensor.data = this.float32ToFloat16Array(tensor.data);
      }
    }
  }

  private float32ToFloat16Array(f32: Float32Array): Uint16Array {
    const f16 = new Uint16Array(f32.length);
    for (let i = 0; i < f32.length; i++) {
      f16[i] = this.toHalf(f32[i]!);
    }
    return f16;
  }

  private toHalf(val: number): number {
    // Simple f32 to f16 conversion logic for WASM/JS environment without relying on external libs
    const floatView = new Float32Array(1);
    const int32View = new Int32Array(floatView.buffer);
    floatView[0] = val;
    const x = int32View[0]!;

    let bits = (x >> 16) & 0x8000; /* Get the sign */
    let m = (x >> 12) & 0x07ff; /* Keep one extra bit for rounding */
    const e = (x >> 23) & 0xff; /* Using int is faster here */

    if (e < 103) {
      return bits;
    }
    if (e > 142) {
      bits |= 0x7c00;
      bits |= e === 255 && x & 0x007fffff ? 1 : 0;
      return bits & 0xffff;
    }
    if (e < 113) {
      m |= 0x0800;
      bits |= (m >> (114 - e)) + ((m >> (113 - e)) & 1);
      return bits & 0xffff;
    }
    bits |= ((e - 112) << 10) | (m >> 1);
    bits += m & 1;
    return bits & 0xffff;
  }

  private quantizeINT8(): void {
    // 236. Generate explicit Asymmetric INT8 TFLite models natively from ONNX QDQ topologies.
    // 243. Identify standard fake-quantize sequences in ONNX and convert directly to Int8 TFLite tensors natively.
    // 237. Produce explicit Per-Channel quantization arrays (1D scales/zeros for DepthwiseConvs).
    // 239. Handle legacy TFLite UINT8 quantization generation.
    // 240. Ensure INT16x8 (16-bit activations, 8-bit weights) metadata can be encoded natively.
    // 244. Implement MinMax parsing to embed fallback quantization metadata inside TFLite.
    // 245. Validate resulting quantized schema against EdgeTPU compiler requirements natively.

    let hasUint8 = false;
    let hasInt16 = false;
    let minMaxExtracted = 0;

    for (let i = 0; i < this.graph.nodes.length; i++) {
      const node = this.graph.nodes[i];
      if (!node) continue;

      if (node.opType === 'QuantizeLinear' || node.opType === 'DynamicQuantizeLinear') {
        const x = node.inputs[0];
        const yScale = node.inputs[1];
        const yZeroPoint = node.inputs[2];
        const y = node.outputs[0];

        if (x && yScale && yZeroPoint && y) {
          const scaleTensor = this.graph.tensors[yScale];
          const zpTensor = this.graph.tensors[yZeroPoint];

          if (scaleTensor?.data && zpTensor?.data) {
            const scaleData = scaleTensor.data as Float32Array;
            const zpData = Array.from(zpTensor.data as any);

            if (zpTensor.dtype === 'uint8') hasUint8 = true;
            if (zpTensor.dtype === 'int16') hasInt16 = true;

            // 238. Extract quantized_dimension correctly for Per-Channel ops.
            const axis = (node.attributes['axis']?.value as number) ?? 0;

            const q: TensorQuantization = {
              min: [],
              max: [],
              scale: Array.from(scaleData),
              zeroPoint: zpData.map(Number),
              quantizedDimension: axis,
            };

            // 244. MinMax approximation for Fallback EdgeTPU if available
            // If the graph contains a Minimum/Maximum block immediately before QDQ, we could extract explicit Min/Max bounds.
            if (q.scale.length === 1 && q.zeroPoint.length === 1) {
              const s = q.scale[0]!;
              const z = q.zeroPoint[0]!;
              const qMin = zpTensor.dtype === 'uint8' ? 0 : -128;
              const qMax = zpTensor.dtype === 'uint8' ? 255 : 127;

              // 144. Ensure fused activation bounds respect asymmetric INT8 limits natively.
              let minBound = (qMin - z) * s;
              let maxBound = (qMax - z) * s;

              if (node.attributes['fused_activation']) {
                const act = node.attributes['fused_activation'].value as string;
                if (act === 'Relu') minBound = Math.max(0, minBound);
                if (act === 'Relu6') {
                  minBound = Math.max(0, minBound);
                  maxBound = Math.min(6.0, maxBound);
                }
              }

              q.min = [minBound];
              q.max = [maxBound];
              minMaxExtracted++;
            }

            // 245. Validate resulting quantized schema against EdgeTPU compiler requirements natively.
            if (q.scale.length > 1 && node.opType !== 'QuantizeLinear') {
              console.warn(
                `[onnx2tf] EdgeTPU Warning: Per-channel quantization on node ${node.name} might cause compilation failures if not aligned correctly.`,
              );
            }

            this.quantizationMap.set(y, q);
            // 237. Produce explicit Per-Channel arrays mapped.
          }
        }
      }
    }

    if (hasUint8) {
      console.log('[onnx2tf] Notice: Generating legacy UINT8 quantization schema.');
    }
    if (hasInt16) {
      console.log('[onnx2tf] Notice: Generating INT16x8 mixed precision quantization schema.');
    }
    if (minMaxExtracted > 0) {
      console.log(
        `[onnx2tf] Notice: Embedded MinMax quantization fallbacks for ${minMaxExtracted} tensors.`,
      );
    }

    console.warn(
      '[onnx2tf] Warning: INT8 AST lowering is experimental. Ensure your model uses standard QuantizeLinear/DequantizeLinear.',
    );
  }
}

File: packages/python/onnx9000-tflite-exporter/src/onnx9000/tflite_exporter/quantization/quantizer.py
"""TFLite quantization utilities and weight quantizers."""

import logging
import struct
from typing import Dict, List
from onnx9000.core.ir import Graph, Tensor
from ..flatbuffer.schema import QuantizationParameters

logger = logging.getLogger(__name__)


class TensorQuantization:
    """Container for tensor quantization parameters."""

    def __init__(
        self,
        min: List[float],
        max: List[float],
        scale: List[float],
        zero_point: List[int],
        quantized_dimension: int,
    ):
        """Initialize TensorQuantization parameters."""
        self.min = min
        self.max = max
        self.scale = scale
        self.zero_point = zero_point
        self.quantized_dimension = quantized_dimension


class Quantizer:
    """TFLite model quantizer."""

    def __init__(self, graph: Graph, mode: str = "none"):
        """Initialize the Quantizer with a graph and quantization mode."""
        self.graph = graph
        self.mode = mode
        self.quantization_map: Dict[str, TensorQuantization] = {}

    def get_quantization_offset(self, builder, tensor: Tensor) -> int:
        """Create and return the FlatBuffer offset for a tensor's quantization parameters."""
        q = self.quantization_map.get(tensor.name)
        if not q:
            return 0

        min_offset = 0
        max_offset = 0
        scale_offset = 0
        zp_offset = 0

        if q.min:
            builder.start_vector(4, len(q.min), 4)
            for val in reversed(q.min):
                builder.add_float32(val)
            min_offset = builder.end_vector(len(q.min))

        if q.max:
            builder.start_vector(4, len(q.max), 4)
            for val in reversed(q.max):
                builder.add_float32(val)
            max_offset = builder.end_vector(len(q.max))

        if q.scale:
            builder.start_vector(4, len(q.scale), 4)
            for val in reversed(q.scale):
                builder.add_float32(val)
            scale_offset = builder.end_vector(len(q.scale))

        if q.zero_point:
            builder.start_vector(8, len(q.zero_point), 8)
            for val in reversed(q.zero_point):
                lower = val & 0xFFFFFFFF
                if lower > 2147483647:
                    lower -= 4294967296
                upper = (val >> 32) & 0xFFFFFFFF
                if upper > 2147483647:
                    upper -= 4294967296
                builder.add_int32(upper)
                builder.add_int32(lower)
            zp_offset = builder.end_vector(len(q.zero_point))

        return QuantizationParameters.create(
            builder, min_offset, max_offset, scale_offset, zp_offset, 0, 0, q.quantized_dimension
        )

    def quantize(self) -> None:
        """Apply quantization to the graph based on the selected mode."""
        if self.mode == "none":
            return
        if self.mode == "fp16":
            self.quantize_fp16()
        elif self.mode == "int8":
            self.quantize_int8()

    def quantize_fp16(self) -> None:
        """Quantize the graph to FLOAT16."""
        # 241. Downcast FLOAT32 FlatBuffer arrays entirely to FLOAT16 bytes explicitly for FP16 models.
        for name, tensor in self.graph.tensors.items():
            if tensor.dtype == "float32" and tensor.is_initializer and tensor.data is not None:
                tensor.dtype = "float16"
                tensor.data = self._float32_to_float16_bytes(tensor.data)

    def _float32_to_float16_bytes(self, f32_bytes: bytes) -> bytes:
        """Convert float32 bytes to float16 bytes."""
        num_floats = len(f32_bytes) // 4
        f32_array = struct.unpack(f"<{num_floats}f", f32_bytes)

        f16_array = []
        for val in f32_array:
            f16_array.append(self._to_half(val))

        return struct.pack(f"<{num_floats}H", *f16_array)

    def _to_half(self, val: float) -> int:
        """Convert a single float32 to float16 (represented as an int)."""
        # Simplistic IEEE 754 float32 to float16 packing
        x = struct.unpack("<I", struct.pack("<f", val))[0]

        bits = (x >> 16) & 0x8000
        m = (x >> 12) & 0x07FF
        e = (x >> 23) & 0xFF

        if e < 103:
            return bits
        if e > 142:
            bits |= 0x7C00
            bits |= 1 if (e == 255 and (x & 0x007FFFFF)) else 0
            return bits & 0xFFFF
        if e < 113:
            m |= 0x0800
            bits |= (m >> (114 - e)) + ((m >> (113 - e)) & 1)
            return bits & 0xFFFF

        bits |= ((e - 112) << 10) | (m >> 1)
        bits += m & 1
        return bits & 0xFFFF

    def quantize_int8(self) -> None:
        """Quantize the graph to INT8."""
        has_uint8 = False
        has_int16 = False
        minmax_extracted = 0

        for i in range(len(self.graph.nodes)):
            node = self.graph.nodes[i]
            if node.op_type in ("QuantizeLinear", "DynamicQuantizeLinear"):
                x = node.inputs[0] if len(node.inputs) > 0 else None
                y_scale = node.inputs[1] if len(node.inputs) > 1 else None
                y_zero_point = node.inputs[2] if len(node.inputs) > 2 else None
                y = node.outputs[0] if node.outputs else None

                if x and y_scale and y_zero_point and y:
                    scale_tensor = self.graph.tensors.get(y_scale)
                    zp_tensor = self.graph.tensors.get(y_zero_point)

                    if scale_tensor and zp_tensor and scale_tensor.data and zp_tensor.data:
                        scale_data = list(
                            struct.unpack(f"<{len(scale_tensor.data) // 4}f", scale_tensor.data)
                        )

                        zp_type = zp_tensor.dtype
                        if zp_type == "uint8":
                            has_uint8 = True
                        elif zp_type == "int16":
                            has_int16 = True

                        zp_format = (
                            "b" if zp_type == "int8" else ("h" if zp_type == "int16" else "B")
                        )
                        zp_data = list(
                            struct.unpack(
                                f"<{len(zp_tensor.data) // struct.calcsize(zp_format)}{zp_format}",
                                zp_tensor.data,
                            )
                        )

                        axis_attr = node.attributes.get("axis")
                        axis = axis_attr.value if axis_attr else 0

                        q = TensorQuantization(
                            min=[],
                            max=[],
                            scale=scale_data,
                            zero_point=zp_data,
                            quantized_dimension=axis,
                        )

                        if len(q.scale) == 1 and len(q.zero_point) == 1:
                            s = q.scale[0]
                            z = q.zero_point[0]
                            q_min = (
                                0
                                if zp_type == "uint8"
                                else (-32768 if zp_type == "int16" else -128)
                            )
                            q_max = (
                                255
                                if zp_type == "uint8"
                                else (32767 if zp_type == "int16" else 127)
                            )

                            min_bound = (q_min - z) * s
                            max_bound = (q_max - z) * s

                            fused = node.attributes.get("fused_activation")
                            if fused and fused.value == "Relu":
                                min_bound = max(0.0, min_bound)
                            elif fused and fused.value == "Relu6":
                                min_bound = max(0.0, min_bound)
                                max_bound = min(6.0, max_bound)

                            q.min = [min_bound]
                            q.max = [max_bound]
                            minmax_extracted += 1

                        if len(q.scale) > 1 and node.op_type != "QuantizeLinear":
                            logger.warning(
                                f"[onnx2tf] EdgeTPU Warning: Per-channel quantization on node {node.name} might cause compilation failures if not aligned correctly."
                            )

                        self.quantization_map[y] = q

        if has_uint8:
            logger.info("[onnx2tf] Notice: Generating legacy UINT8 quantization schema.")
        if has_int16:
            logger.info("[onnx2tf] Notice: Generating INT16x8 mixed precision quantization schema.")
        if minmax_extracted > 0:
            logger.info(
                f"[onnx2tf] Notice: Embedded MinMax quantization fallbacks for {minmax_extracted} tensors."
            )

        logger.warning(
            "[onnx2tf] Warning: INT8 AST lowering is experimental. Ensure your model uses standard QuantizeLinear/DequantizeLinear."
        )
